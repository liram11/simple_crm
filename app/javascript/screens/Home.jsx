import React, { useEffect, useState } from "react";
import lodash from "lodash";

import { Input } from "../ui/Input";
import { Error } from "../ui/Error";
import { CompaniesTable } from "../ui/CompaniesTable";

import { fetchCompanies } from "../utils/api";


export default function Home() {
  // List of fetched companies
  const [companies, setCompanies] = useState([]);

  // Table filters
  const [companyName, setCompanyName] = useState("");
  const [industry, setIndustry] = useState("");
  const [minEmployee, setMinEmployee] = useState("");
  const [minDealAmount, setMinDealAmount] = useState("");

  // Last error
  const [error, setError] = useState("");

  // Pagination
  const [page, setPage] = useState(1);
  const [totalPages, setTotalPages] = useState(0);

  // Fetch companies from API
  useEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    const debouncedFetchCompanies = lodash.debounce(() => {
      setError("");

      fetchCompanies({
        companyName,
        industry,
        minEmployee,
        minDealAmount,
        page
      }, { signal })
        .then((res) => {
          setCompanies(res.companies);
          setTotalPages(res.meta.total_pages);
        })
        .catch((e) => {
          if (e.name === "AbortError") { return; }
          setError(e.message);
        });
    }, 400);

    debouncedFetchCompanies();

    return () => {
      controller.abort();
    };
  }, [companyName, industry, minEmployee, minDealAmount, page]);

  return (
    <div className="vw-100 primary-color d-flex align-items-center justify-content-center">
      <div className="jumbotron vw-100 jumbotron-fluid bg-transparent">
        <div className="container secondary-color">
          <h1 className="display-4">Companies</h1>

          <div className="filters">
            <Input id="company-name" label="Company Name" value={companyName} onChange={setCompanyName} />
            <Input id="industry" label="Industry" value={industry} onChange={setIndustry} />
            <Input id="min-employee" type="number" label="Minimum Employee Count" value={minEmployee} onChange={setMinEmployee} />
            <Input id="min-amount" type="number" label="Minimum Deal Amount" value={minDealAmount} onChange={setMinDealAmount} />
          </div>

          <Error message={error} />

          <CompaniesTable
            companies={companies}
            page={page}
            onPageChange={setPage}
            totalPages={totalPages}
          />
        </div>
      </div>
    </div>
  );
}
