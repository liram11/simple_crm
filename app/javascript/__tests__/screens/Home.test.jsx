/**
 * @jest-environment jsdom
 */

import React from "react";
import renderer from "react-test-renderer";
import Home from "../../screens/Home";

// eslint-disable-next-line no-undef
it("renders correctly", () => {
  const tree = renderer
    .create(<Home id="some_id" value="some value" label="Some Label" onChange={() => {}}/>)
    .toJSON();
    // eslint-disable-next-line no-undef
  expect(tree).toMatchSnapshot();
});
