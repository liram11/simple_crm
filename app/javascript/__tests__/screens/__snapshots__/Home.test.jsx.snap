// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="vw-100 primary-color d-flex align-items-center justify-content-center"
>
  <div
    className="jumbotron jumbotron-fluid bg-transparent"
  >
    <div
      className="container secondary-color"
    >
      <h1
        className="display-4"
      >
        Companies
      </h1>
      <div
        className="filters"
      >
        <label
          htmlFor="company-name"
        >
          Company Name
        </label>
        <div
          className="input-group mb-3"
        >
          <input
            className="form-control"
            id="company-name"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <label
          htmlFor="industry"
        >
          Industry
        </label>
        <div
          className="input-group mb-3"
        >
          <input
            className="form-control"
            id="industry"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <label
          htmlFor="min-employee"
        >
          Minimum Employee Count
        </label>
        <div
          className="input-group mb-3"
        >
          <input
            className="form-control"
            id="min-employee"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <label
          htmlFor="min-amount"
        >
          Minimum Deal Amount
        </label>
        <div
          className="input-group mb-3"
        >
          <input
            className="form-control"
            id="min-amount"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <table
        className="table"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              Name
            </th>
            <th
              scope="col"
            >
              Industry
            </th>
            <th
              scope="col"
            >
              Employee Count
            </th>
            <th
              scope="col"
            >
              Total Deal Amount
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</div>
`;
